### DISCUSSION TITLE
401 invalid client

### DISCUSSION THREAD

**Discussion by just-dev-creator** [TheDrone7/disco-oauth#39] [Q&A] [answered]
To be clear: This is my first time using this package. I apologize for an maybe "nooby" error on my side.

To have a basis for this message, here are the (maybe) affected code snippets:
```javascript
const client = new Client("835953723125137449", process.env.CLIENT_SECRET)
client.setScopes(['guilds'])
client.setRedirect("http://localtest.me:8080/v1/dc-login")
```
```javascript
router.get("/dc-login", async (req, res) => {
    if (req.query.code == null) {
        res.writeHead(400, "No discord auth code specified")
        res.end()
    }
    const code = req.query.code
    const key = client.getAccess(req.query.code)
})
```
If I just straight up execute this code and authorize the application, I get the following error:
```
[nodemon] starting `node index.js`
/home/justus/WebstormProjects/etb-item-recovery-api/etb-item-recovery-api/node_modules/disco-oauth/lib/client.js:154
        } else reject(new APIError(response));
                      ^

APIError: [28.4.2021, 19:51:38] :: Discord API Error encountered.
undefined

    at /home/justus/WebstormProjects/etb-item-recovery-api/etb-item-recovery-api/node_modules/disco-oauth/lib/client.js:154:23
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
[nodemon] app crashed - waiting for file changes before starting...
```
Doing some further investigation (printing out  the response code and body in client.js:146), I got the error code 401 and the error `invalid_client` back from the api. 
What I can confirm:

- [X] The redirect URI is correct (just to be shure, copy-pasted from the dev portal: `http://localtest.me:8080/v1/dc-login`)
- [X] Client id and secret are correct (censorded, working http request that i've done with the same creds
```http
POST https://discord.com/api/oauth2/token
Content-Type: application/x-www-form-urlencoded


client_id=835953723125137449&client_secret=**SECRET**&grant_type=authorization_code&code=**SECRET**&redirect_uri=http%3A%2F%2Flocaltest.me%3A8080%2Fv1%2Fdc-login
```
I am really thankful for anyone who can help me out. 

Justus

**Comment by TheDrone7** [comment, marked as answer]
The issue here is in your route, I believe. Even when you use `res.end()` inside the if block. The rest of the code (outside the if block) still gets executed, causing the problem. You may want to put the rest inside an else block to only execute it if the code actually exists.

Also, instead of `if (req.query.code == null)`, simply use `if (req.query.code)`. `null` in JS is a valid value and else block will get executed even if it is `undefined`.

**Reply by just-dev-creator** [reply]
...and thats the experience of developing in js coming from java.
